### PR TITLE
Add missing java loads

### PR DESCRIPTION
**Pull request by ted-xie**

### CODE REVIEW

File: java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package(default_visibility = ["//visibility:public"])
 
+load("@rules_java//java:defs.bzl", "java_library")
```
Comment (Wyverald) [inline java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel:3]: let's load from https://github.com/bazelbuild/rules_java/blob/master/java/java_library.bzl instead

Comment (ted-xie) [reply java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel:3]: I'm fine with either way, but is there official guidance on which path is more official? Loading from defs.bzl is more widespread in cs.opensource.google https://cs.opensource.google/search?q=load.*java_library&sq=

Comment (ted-xie) [reply java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel:3]: A similar search in bazel-contrib: https://github.com/search?q=org%3Abazel-contrib+%22load%28%5C%22%40rules_java%22&type=code

Comment (Wyverald) [reply java/src/main/java/com/google/devtools/build/lib/worker/BUILD.bazel:3]: defs.bzl is the old convention; the per-rule .bzl reduces transitive loads. See [this doc comment](https://docs.google.com/document/d/1L1JFgjpZ7SrBinb24DC_5nTIELeYDacikcme-YcA7xs/edit?disco=AAAA_GneyNw) (and the whole doc actually)

**Comment by ted-xie** [comment]
Changing the java_library load path seems to have broken a bunch of tests, but I can't open the Buildkite results. I'm logged into Buildkite, though.

image

**Comment by Wyverald** [comment]
Sorry about that -- I forgot to set the proper ACLs. I just added the Bazel team on BuildKite and you should be able to see the errors now. (We need to update .bazelversion to at least 7.3.2.)

Thanks for helping me work through the setup :)

**Comment by Wyverald** [comment]
```
  | (20:05:43) ERROR: Traceback (most recent call last):
  | File "C:/b/ivwz3ksz/external/rules_java~/java/common/rules/android_lint.bzl", line 142, column 31, in <toplevel>
  | android_lint_subrule = subrule(
  | Error in subrule: file '@@rules_java~//java/common/rules:android_lint.bzl' cannot use private API
  | (20:05:43) WARNING: Target pattern parsing failed.
  | (20:05:43) ERROR: Skipping '//...': error loading package under directory '': error loading package 'src/main/java/com/google/devtools/build/lib/worker': at C:/b/ivwz3ksz/external/rules_java~/java/java_library.bzl:16:6: at C:/b/ivwz3ksz/external/rules_java~/java/bazel/rules/bazel_java_library.bzl:21:6: initialization of module 'java/common/rules/android_lint.bzl' failed
  | (20:05:43) ERROR: error loading package under directory '': error loading package 'src/main/java/com/google/devtools/build/lib/worker': at C:/b/ivwz3ksz/external/rules_java~/java/java_library.bzl:16:6: at C:/b/ivwz3ksz/external/rules_java~/java/bazel/rules/bazel_java_library.bzl:21:6: initialization of module 'java/common/rules/android_lint.bzl' failed
```

hmm. no idea what this is. Maybe try Bazel 7.4.0?

cc @hvadehra

**Comment by ted-xie** [comment]
Bazel 7.4 had the same failure, let's get hvadehra's opinion on this.

**Comment by ted-xie** [comment]
OK, I also just realized something weird: Loading from `@rules_java//java:defs.bzl` passed BazelCI (see the first commit in this PR), but changing the load path caused this strange subrule error.

**Comment by Wyverald** [comment]
The first commit did not run CI actually (click on the green tick and you can see that only the CLA check ran). I hadn't set the BuildKite pipeline to trigger on forks yet (another thing to note down in my notebook...)

**Comment by hvadehra** [comment]
Do you actually need/want rules_java v8.1.0? In general, for Bazel version N, prefer rules_java version N (latest is [7.12.2](https://github.com/bazelbuild/rules_java/releases/tag/7.12.2))

It is possible to use rules_java 8 with Bazel 7, but will need enabling the `--experimental_rule_extension_api` flag (enabled by default in Bazel 8).

**Comment by ted-xie** [comment]
> for Bazel version N, prefer rules_java version N 

What do you suggest for rule or util repositories that need to support multiple major versions? I'm thinking particularly about rules_android, which has to support both Bazel 7 and 8, and so do its dependencies.

**Comment by hvadehra** [comment]
For such cases, I thought one would just use the Bazel builtin version? i.e. skip adding an explicit `bazel_dep` for rules_java

**Comment by ted-xie** [comment]
>  i.e. skip adding an explicit bazel_dep for rules_java

I tried this in my local repo (together with `--override_module=bazel_worker_java=/path/to/bazel_worker_api/java`) and it failed with this error:

```
ERROR: error loading package '@@bazel_worker_java+//src/main/java/com/google/devtools/build/lib/worker': Unable to find package for @@[unknown repo 'rules_java' requested from @@bazel_worker_java+]//java:java_library.bzl: The repository '@@[unknown repo 'rules_java' requested from 
@@bazel_worker_java+]' could not be resolved: No repository visible as '@rules_java' from repository '@@bazel_worker_java+'.                                                                                                                                                              
ERROR: /usr/local/google/home/tedx/work/rules_android/src/tools/java/com/google/devtools/build/android/r8/BUILD:67:13: error loading package '@@bazel_worker_java+//src/main/java/com/google/devtools/build/lib/worker': Unable to find package for @@[unknown repo 'rules_java' requested
 from @@bazel_worker_java+]//java:java_library.bzl: The repository '@@[unknown repo 'rules_java' requested from @@bazel_worker_java+]' could not be resolved: No repository visible as '@rules_java' from repository '@@bazel_worker_java+'. and referenced by '//src/tools/java/com/googl
e/devtools/build/android/r8:dexerdeps'       
```

So it seems like we do need to explicitly declare rules_java as a bazel_dep for this module.

**Comment by Wyverald** [comment]
what about the autoloads stuff, though? Actually, why do we need this PR at all? I thought it was just for future compatibility. If we don't add a `load` for `java_library`, it should just be autoloaded on Bazel 8, right?

**Comment by ted-xie** [comment]
Honestly I am not really sure exactly how the autoloads work -- specifically, when certain repositories are special-cased. As it stands currently, when the main repository is rules_android, this module's java rule instances are not autolaoded.

I put together a minimal repro here: https://github.com/ted-xie/autoloads_with_dummy_rules_android

FYI @comius

**Review by Wyverald — APPROVED** [review]

**Comment by hvadehra** [comment]
Update, rules_java 8.3.0 is available that does not suffer from this problem.

**Pull request merged**